File: api/routes/intelligence.py
"""Lead Intelligence endpoints (Prompt 50, §43).

Aggregation read-model over existing engines. GET is open (deterministic, cached — no
provider call forced). The refresh POST re-runs the bounded AI analysis and is
role-guarded + rate-limited so provider calls never happen on every page load (§44/§45).
No new URL prefix — follows the codebase convention (routes declare full paths).
"""

from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from api.dependencies import get_session
from api.schemas import (
    CompanyProfileResponse,
    LeadIntelligenceResponse,
    LeadSourcesResponse,
    OpportunityViewResponse,
    POCIntelligenceResponse,
    POCResponse,
    ProfileHighlightsResponse,
    SourceRefResponse,
    StakeholderRoleResponse,
)
from api.security import rate_limit, require_role
from config.exceptions import NotFoundError
from database.models import Lead, UserRole
from intelligence.lead_intelligence import (
    LeadIntelligence,
    POCIntelligence,
    build_lead_intelligence,
    lead_sources,
)

router = APIRouter(tags=["intelligence"])

_refresh_role = require_role(UserRole.SALES)
_refresh_limit = rate_limit("intelligence_refresh", 20)


def _get_lead(session: Session, lead_id: int) -> Lead:
    lead = session.get(Lead, lead_id)
    if lead is None:
        raise NotFoundError(f"Lead {lead_id} not found.")
    return lead


def _poc_response(pi: POCIntelligence) -> POCIntelligenceResponse:
    return POCIntelligenceResponse(
        poc=POCResponse.model_validate(pi.poc), poc_status=pi.poc_status,
        role_match_score=pi.role_match_score, contact_trust_score=pi.contact_trust_score,
        contact_trust_status=pi.contact_trust_status, is_primary=pi.is_primary)


def _to_response(intel: LeadIntelligence) -> LeadIntelligenceResponse:
    roles = [StakeholderRoleResponse(role=r.role, role_category=r.role_category,
                                     decision_maker_type=r.decision_maker_type,
                                     relevance_score=r.relevance_score, reason=r.reason,
                                     is_primary=r.is_primary) for r in intel.recommended_roles]
    ov = intel.opportunity
    return LeadIntelligenceResponse(
        lead_id=intel.lead_id, company_id=intel.company_id, company_name=intel.company_name,
        lead_score=intel.lead_score, lead_priority=intel.lead_priority,
        data_trust=intel.data_trust, contact_trust=intel.contact_trust,
        signal_summary=intel.signal_summary,
        company_profile=CompanyProfileResponse(**intel.company_profile.as_dict()),
        opportunity=OpportunityViewResponse(
            opportunity_type=ov.opportunity_type, label=ov.label, staffing_need=ov.staffing_need,
            estimated_team=ov.estimated_team, urgency=ov.urgency, technologies=ov.technologies,
            signals=ov.signals),
        recommended_roles=roles, recommendation_confidence=intel.recommendation_confidence,
        primary_poc=_poc_response(intel.primary_poc) if intel.primary_poc else None,
        secondary_pocs=[_poc_response(p) for p in intel.secondary_pocs],
        ai_highlights=ProfileHighlightsResponse(**intel.ai_highlights.as_dict()),
        sources=[SourceRefResponse(**s.as_dict()) for s in intel.sources],
        freshness_score=intel.freshness_score, freshness_status=intel.freshness_status,
        sales_action=intel.sales_action, generated_at=intel.generated_at)


@router.get("/leads/{lead_id}/intelligence", response_model=LeadIntelligenceResponse,
            summary="Aggregated lead intelligence (company + signal + POC + trust + AI highlights)")
def get_lead_intelligence(lead_id: int, session: Session = Depends(get_session)) -> LeadIntelligenceResponse:
    lead = _get_lead(session, lead_id)
    return _to_response(build_lead_intelligence(session, lead, refresh=False))


@router.get("/leads/{lead_id}/sources", response_model=LeadSourcesResponse,
            summary="All contributing sources for a lead (field-level provenance)")
def get_lead_sources(lead_id: int, session: Session = Depends(get_session)) -> LeadSourcesResponse:
    lead = _get_lead(session, lead_id)
    from enrichment.contactout_poc import resolve_company_for_lead
    company, company_name, _ = resolve_company_for_lead(session, lead)
    return LeadSourcesResponse(
        lead_id=lead.id, company_id=(company.id if company else None), company_name=company_name,
        sources=[SourceRefResponse(**s.as_dict()) for s in lead_sources(session, lead)])


@router.post("/leads/{lead_id}/intelligence/refresh", response_model=LeadIntelligenceResponse,
             summary="Recompute lead intelligence (bounded AI re-analysis)")
def refresh_lead_intelligence(lead_id: int, session: Session = Depends(get_session),
                              role=Depends(_refresh_role), _rl=Depends(_refresh_limit)
                              ) -> LeadIntelligenceResponse:
    lead = _get_lead(session, lead_id)
    return _to_response(build_lead_intelligence(session, lead, refresh=True))

File: frontend/src/services/intelligence.ts
import { api } from '@/services/api';
import type { POC, RecommendedRole } from '@/services/contactOut';

/** Lead Intelligence aggregation (Prompt 50). Read-model over existing engines —
 *  every value is source-backed; nothing is fabricated. */

export type PocStatus =
  | 'VERIFIED'
  | 'LIKELY'
  | 'RECOMMENDED_ROLE_ONLY'
  | 'STALE'
  | 'FORMER'
  | 'UNVERIFIED';

export type HighlightTrust = 'HIGH' | 'MEDIUM' | 'LOW';
export type FreshnessStatus = 'FRESH' | 'AGING' | 'STALE' | 'UNKNOWN';

export interface SourceRef {
  field: string;
  value: string | null;
  source: string;
  source_type: string | null;
  source_url: string | null;
  source_record_id: string | null;
  trust_score: number;
  verification_status: string | null;
  evidence: string | null;
  retrieved_at: string | null;
  last_verified_at: string | null;
}

export interface ProfileHighlight {
  highlight_title: string;
  highlight_text: string;
  trust_level: HighlightTrust;
  supporting_signal_ids: number[];
  supporting_source_ids: string[];
  ai_generated: boolean;
}

export interface ProfileHighlights {
  highlights: ProfileHighlight[];
  ai_available: boolean;
  ai_insight: string | null;
  model_version: string | null;
  generated_at: string | null;
}

export interface CompanyProfile {
  company_name: string | null;
  industry: string | null;
  india_presence: string;
  india_entity_type: string | null;
  website: string | null;
  primary_location: string | null;
  registered_location: string | null;
  career_site: string | null;
  technology_focus: string[];
  current_hiring_signal: string | null;
  opportunity: string | null;
  data_trust: number;
  headline: string;
}

export interface OpportunityViewData {
  opportunity_type: string;
  label: string;
  staffing_need: string;
  estimated_team: number | null;
  urgency: string;
  technologies: string[];
  signals: string[];
}

export interface POCIntelligence {
  poc: POC;
  poc_status: PocStatus;
  role_match_score: number;
  contact_trust_score: number;
  contact_trust_status: string | null;
  is_primary: boolean;
}

export interface LeadIntelligence {
  lead_id: number;
  company_id: number | null;
  company_name: string | null;
  lead_score: number;
  lead_priority: string | null;
  data_trust: number;
  contact_trust: number;
  signal_summary: string;
  company_profile: CompanyProfile;
  opportunity: OpportunityViewData;
  recommended_roles: RecommendedRole[];
  recommendation_confidence: number;
  primary_poc: POCIntelligence | null;
  secondary_pocs: POCIntelligence[];
  ai_highlights: ProfileHighlights;
  sources: SourceRef[];
  freshness_score: number;
  freshness_status: FreshnessStatus;
  sales_action: string;
  generated_at: string | null;
}

export interface LeadSources {
  lead_id: number;
  company_id: number | null;
  company_name: string | null;
  sources: SourceRef[];
}

export async function fetchLeadIntelligence(leadId: number, signal?: AbortSignal): Promise<LeadIntelligence> {
  const { data } = await api.get<LeadIntelligence>(`/leads/${leadId}/intelligence`, { signal });
  return data;
}

export async function fetchLeadSources(leadId: number, signal?: AbortSignal): Promise<LeadSources> {
  const { data } = await api.get<LeadSources>(`/leads/${leadId}/sources`, { signal });
  return data;
}

export async function refreshLeadIntelligence(leadId: number): Promise<LeadIntelligence> {
  const { data } = await api.post<LeadIntelligence>(`/leads/${leadId}/intelligence/refresh`);
  return data;
}

// --- Display helpers (label + className together, never color alone) --------- //
const EMERALD = 'bg-emerald-100 text-emerald-700 dark:bg-emerald-500/15 dark:text-emerald-300';
const BLUE = 'bg-blue-100 text-blue-700 dark:bg-blue-500/15 dark:text-blue-300';
const AMBER = 'bg-amber-100 text-amber-700 dark:bg-amber-500/15 dark:text-amber-300';
const SLATE = 'bg-slate-100 text-slate-600 dark:bg-slate-700/40 dark:text-slate-300';
const ROSE = 'bg-rose-100 text-rose-700 dark:bg-rose-500/15 dark:text-rose-300';

export function pocStatusDisplay(status: PocStatus): { label: string; className: string } {
  switch (status) {
    case 'VERIFIED':
      return { label: 'Verified', className: EMERALD };
    case 'LIKELY':
      return { label: 'Likely', className: BLUE };
    case 'STALE':
      return { label: 'Stale', className: AMBER };
    case 'FORMER':
      return { label: 'Former', className: SLATE };
    case 'RECOMMENDED_ROLE_ONLY':
      return { label: 'Recommended role only', className: SLATE };
    default:
      return { label: 'Unverified', className: AMBER };
  }
}

export function trustLevelDisplay(level: HighlightTrust): { label: string; className: string } {
  if (level === 'HIGH') return { label: 'High', className: EMERALD };
  if (level === 'MEDIUM') return { label: 'Medium', className: BLUE };
  return { label: 'Low', className: AMBER };
}

export function freshnessDisplay(status: FreshnessStatus): { label: string; className: string } {
  if (status === 'FRESH') return { label: 'Fresh', className: EMERALD };
  if (status === 'AGING') return { label: 'Aging', className: AMBER };
  if (status === 'STALE') return { label: 'Stale', className: ROSE };
  return { label: 'Unknown', className: SLATE };
}
